---
title: "DESeq2 Workflow"
output: html_document
author: Christof Seiler
date: October, 2016
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Goal

Establish DESeq2 workflow starting from BAM files generated by tophat2. 
To get the BAM files first install all necessary software and submit aligment script to sherlock. 

1. Run ``bash Installation.bash`` in your ``$HOME`` folder 
2. Run ``tophat2`` aligment using ``bash Alignment.bash`` in your ``$SCRATCH`` folder

This is based on [DESeq2 vignette](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf) and entire [DESeq2 workflow documentation](http://www.bioconductor.org/help/workflows/rnaseqGene/).

## Workfow

Load packages.

```{r message=FALSE}
#source("https://bioconductor.org/biocLite.R")
#biocLite("Rsamtools")
#biocLite("GenomicFeatures")
library("Rsamtools")
library("GenomicFeatures")
library("BiocParallel")
n_cores = as.integer(Sys.getenv("SLURM_NTASKS"))
n_cores
register(MulticoreParam(workers = n_cores))
```

Import BAM file into R.

```{r}
csvfile = file.path("sample_table.csv")
sample_table = read.csv(csvfile,row.names=1)
file_names = paste0(sample_table$name,".bam")
bamfiles = BamFileList(file_names, yieldSize=2000000)
seqinfo(bamfiles[1])
```

Define gene model (this is downloaded during ``Installation.bash`` automatically).

```{r}
gene_model_path = Sys.getenv("GENE_MODEL")
gtffile = file.path(gene_model_path,"Homo_sapiens.GRCh38.86.gtf")
txdb = makeTxDbFromGFF(gtffile, format="gtf", circ_seqs=character())
txdb
```

List of all the exons grouped by gene.

```{r}
ebg = exonsBy(txdb, by="gene")
ebg
```

Read counting step.

```{r}
se = summarizeOverlaps(features=ebg, reads=bamfiles,
                       mode="Union",
                       singleEnd=FALSE,
                       ignore.strand=TRUE,
                       fragments=TRUE )
se
rowRanges(se)
str(metadata(rowRanges(se)))
ta(se) = DataFrame(sampleTable)
ta(se)
```

This completes the counting step, we have counted the fragments which overlap the genes in the gene model we specified. Now we can feed this into many different kind of linear models in Bioconductor, one option is to use the ``DESeq2`` package. Before we continue, we save the *SummarizedExperiment* so that we don't need to rerun the counting step again.

```{r}
save(se,file = "se.Rdata")
```
