---
title: "Gene Expression Network Analysis"
author: Christof Seiler
date: October, 2016
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This Rmd file contains a minimal network analysis pipeline using the packages ``STRINGdb`` and ``BioNet``.

## Install Packages

Install necessary packages from bioconductor repository. Run this code only once to install packages.

```{r eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("STRINGdb")
biocLite("BioNet")
biocLite("DLBCL")
install.packages(c("stringr","igraph","visNetwork"))
```

Load packages.

```{r warning=FALSE,message=FALSE}
library(STRINGdb)
library(BioNet)
library(DLBCL)
library(stringr)
library(igraph)
library(visNetwork)
```

## Load Data

Remove genes with NA values. From the [DESeq2 documentation](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf). Some values in the results table can be set to NA for one of the following reasons:

1. If within a row, all samples have zero counts, the baseMean column will be zero, and the log2 fold change estimates, p-value and adjusted p-value will all be set to NA.
2. If a row contains a sample with an extreme count outlier then the p-value and adjusted p-value will be set to NA. These outlier counts are detected by Cook's distance.
3. If a row is filtered by automatic independent filtering, for having a low mean normalized count, then only the adjusted p-value will be set to NA.

```{r}
diff_exp = read.csv("treatment_H1N1_H3N2_results.csv")
dim(diff_exp)
diff_exp = na.omit(diff_exp)
dim(diff_exp)
```

## Download Gene Network

Download proteins for human species (code is 9606).

Consider interaciton that are 0.9 confidence. From the STRING website: 

In STRING, each protein-protein interaction is annotated with one or more 'scores'. Importantly, these scores do not indicate the strength or the specificity of the interaction. Instead, they are indicators of confidence, i.e. how likely STRING judges an interaction to be true, given the available evidence. All scores rank from 0 to 1, with 1 being the highest possible confidence. A score of 0.5 would indicate that roughly every second interaction might be erroneous (i.e., a false positive).

```{r}
string_db = STRINGdb$new(version="10", species=9606, score_threshold=900, input_directory="")
```

Check how many proteins are in the database.

```{r}
string_proteins = string_db$get_proteins()
dim(string_proteins)
```

Map gene names to identifiers used in the database.

```{r}
mapped = string_db$map(diff_exp, "X", removeUnmappedRows = FALSE)
interactions = string_db$get_interactions(mapped$STRING_id)
interactions = data.frame(from=interactions$from,to=interactions$to,combined_score=interactions$combined_score)
dim(interactions)
head(interactions)
```

## Find Subgraph

Fit a Beta-Uniform model.

```{r fig.width=12}
pval = mapped$pvalue
names(pval) = mapped$STRING_id
fb = fitBumModel(pval)
fb
```

Set the ``fdr`` parameter which can be interpreted as the FDR of the subgraph. Smaller values will produce a smaller maximum subgraph. You should try a few values (e.g. 0.05, 0.01, 0.001, ...) to obtain a reasonable small subgraph that permits biological interpretation.

First we make convert the interaction table into an igraph object and make the nodes names human readible.

```{r}
network = graph_from_data_frame(interactions)
```

Then we search for the optimal subgraph.

```{r}
scores = scoreNodes(network, fb, fdr=0.01)
module = runFastHeinz(network, scores)
module
```

Here the list of nodes within that module.

```{r}
mapped[which(mapped$STRING_id %in% names(V(module))),]
```

Differential expression is coloured in red (upregulated) and green (downregulated). Scores are represented with shapes: rectangles are negative and circles are positive scores.

```{r fig.width=12, fig.height=12}
if(length(V(module)) > 0) {
  par(mfrow=c(1,1))
  logFC = mapped$log2FoldChange
  names(logFC) = mapped$STRING_id
  plotModule(module, diff.expr = logFC)
} else {
  message("no signficant subgraph found")
}
```

Interactive visualization.

```{r fig.width=12, fig.height=12}
if(length(V(module)) > 0) {
  V(module)$label = names(V(module))
  visIgraph(module, idToLabel = FALSE)
} else {
  message("no signficant subgraph found")
}
```

Interactive visualization with physics animation.

```{r fig.width=12, fig.height=12}
if(length(V(module)) > 0) {
  V(module)$label = names(V(module))
  visIgraph(module, idToLabel = FALSE, physics = TRUE)
} else {
  message("no signficant subgraph found")
}
```
