---
title: "Transcription of Influenza Strain"
output: html_document
author: Christof Seiler
date: June, 2017
params:
  type: "sample_table_monocyte_all.csv"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Goal

Count influenza strain gene expression in monocyte co-culture. The two strains are:

* A_California_07_2009
* A_Victoria_361_2011

```{r}
library(Rsamtools)
library(GenomicFeatures)
library(GenomicAlignments)
library(ggplot2)
library(reshape2)
library(stringr)
library(devtools)
theme_set(theme_bw())
strain_colors = NULL
strain_colors$H1N1 = "#619CFF"
strain_colors$H3N2 = "#F8766D"
strain_colors$UI = "#00BA38"
params
virus_names = c("A_California_07_2009","A_Victoria_361_2011")
```

## Prepare Count Matrix

Manually rename sequence levels because there is a mismatch between .fasta and .gff files download from [fludb.org](fludb.org).

```{r}
seqnames_mapping = NULL
seqnames_mapping[[virus_names[1]]] = c('832340' = 'gb:CY121687|Organism:Influenza',
                                       '832372' = "gb:CY121681|Organism:Influenza",
                                       '833612' = "gb:CY121682|Organism:Influenza",
                                       '835512' = "gb:CY121683|Organism:Influenza",
                                       '835525' = "gb:CY121680|Organism:Influenza",
                                       '839114' = "gb:CY121686|Organism:Influenza",
                                       '839546' = "gb:CY121684|Organism:Influenza",
                                       '856776' = "gb:CY121685|Organism:Influenza")
seqnames_mapping[[virus_names[2]]] = c('1185941' = 'gb:KJ942683|Organism:Influenza',
                                       '1185945' = 'gb:KJ942684|Organism:Influenza',
                                       '1185959' = 'gb:KJ942680|Organism:Influenza',
                                       '1185960' = 'gb:KJ942681|Organism:Influenza',
                                       '1185990' = 'gb:KJ942682|Organism:Influenza',
                                       '1185991' = 'gb:KJ942686|Organism:Influenza',
                                       '1185992' = 'gb:KJ942687|Organism:Influenza',
                                       '1186070' = 'gb:KJ942685|Organism:Influenza')
```

Import BAM file into R. Define gene model with downloaded .gff files. Rename. Count.

```{r}
make_count_matrix = function(virus) {
  csvfile = file.path(params$type)
  sample_table = read.csv(csvfile,row.names=1)
  file_names = paste0(sample_table$name,"_",virus,"_STAR.bam")
  bamfiles = BamFileList(file_names, yieldSize=2000000)
  #seqinfo(bamfiles[1])
  gff_file = paste0(virus,"_GffResults.gff")
  txdb = makeTxDbFromGFF(gff_file,format = "gff3")
  ebg = exonsBy(txdb, by="tx", use.names=TRUE)
  ebg_renamed = renameSeqlevels(ebg, seqnames_mapping[[virus]])
  #seqnames(ebg)
  #seqnames(ebg_renamed)
  se = summarizeOverlaps(features=ebg_renamed, reads=bamfiles,
                         mode="Union",
                         singleEnd=FALSE,
                         ignore.strand=TRUE,
                         fragments=TRUE)
  colData(se) = DataFrame(sample_table)
  se
}
se_list = lapply(virus_names,make_count_matrix)
se_list
```

## Analyze Count Matrix

Extract count matrix. Need to account for library size. We can access the size factors from our DESeq2 analysis on monocytes. Need to run this file in the same folder as the DESeq2 analysis. Facet by gene. Plot both for quality control.

```{r fig.height=7,fig.width=10}
normalize_counts = function(counts) {
  load("dds.Rdata")
  ids = sapply(colnames(counts),function(sample_name) which(sample_name == names(sizeFactors(dds))))
  counts = counts[,ids]
  counts_nonzero = counts[rowSums(counts) > 0,]
  counts_nonzero_normalized = counts_nonzero/sizeFactors(dds)
  counts_nonzero_normalized
}
p_separate = lapply(1:2,function(i) {
  se = se_list[[i]]
  counts_nonzero_normalized = normalize_counts(assay(se))
  combo = merge(t(counts_nonzero_normalized),colData(se),by="row.names")
  combo_long = melt(combo,
                    id.vars = c("Row.names","name","treatment","harvest","donor"))
  combo_long$variable = factor(combo_long$variable, 
                               levels = sort(levels(combo_long$variable)))
  combo_long$harvest = factor(combo_long$harvest, 
                              levels = c("7HPI","24HPI"))
  ind = which(names(strain_colors) %in% levels(combo_long$treatment))
  strain_colors_sel = unlist(strain_colors[ind])
  ggplot(combo_long, aes(x=harvest, y=value, color = treatment, shape=donor)) +
    geom_point(position=position_jitter(w=0.3,h=0), size=3) + 
    scale_y_log10() +
    facet_wrap(~variable,nrow = 2) +
    ylab("normalized and log10 transformed counts") +
    ggtitle(str_replace_all(string = virus_names[i],pattern = "_",replacement = "/")) +
    scale_color_manual(values=strain_colors_sel)
})
p_separate
ggsave(p_separate[[1]],filename = paste0("Virus_Expression_",virus_names[1],".pdf"),height = 5,width = 9)
ggsave(p_separate[[2]],filename = paste0("Virus_Expression_",virus_names[2],".pdf"),height = 5,width = 9)
```

Compare the two strains.

```{r fig.height=7,fig.width=10}
combo = lapply(1:2,function(i) {
  se = se_list[[i]]
  counts_nonzero_normalized = normalize_counts(assay(se))
  combo = merge(t(counts_nonzero_normalized),colData(se),by="row.names")
  combo$strain = virus_names[i]
  combo
}) %>% do.call(rbind,.)
combo_subset = rbind(subset(combo,treatment == "H1N1" & strain == "A_California_07_2009"),
                     subset(combo,treatment == "H3N2" & strain == "A_Victoria_361_2011"))
combo_subset_long = melt(combo_subset,
                         id.vars = c("Row.names","name","treatment","harvest","donor","strain"))
combo_subset_long$variable = factor(combo_subset_long$variable, 
                                    levels = sort(levels(combo_subset_long$variable)))
combo_subset_long$harvest = factor(combo_subset_long$harvest, 
                                   levels = c("7HPI","24HPI"))
ind = which(names(strain_colors) %in% levels(combo_subset_long$treatment))
strain_colors_sel = unlist(strain_colors[ind])
p_compare = ggplot(combo_subset_long, aes(x=harvest, y=value, col=treatment, shape=donor)) +
  geom_point(position=position_jitter(w=0.3,h=0), size=3) + 
  scale_y_log10() +
  facet_wrap(~variable,nrow = 2) +
  ylab("normalized and log10 transformed counts") +
  scale_color_manual(values=strain_colors_sel)
p_compare
ggsave(p_compare,filename = "Virus_Expression_Comparison.pdf",height = 4,width = 9)
```

## Session Info

```{r}
session_info()
```
